common/nng/socket.cpp

```diff
@@ -39,14 +39,23 @@ Socket::~Socket()
 
 bool Socket::SyncRequired() { return false; }
 
-int Socket::Connect(const std::string &addr, bool retry)
+int Socket::Connect(const std::string &addr, int retry)
 {
     mapf_assert(!addr.empty());
     int rc;
     do {
         rc = nng_dial(sock->sd_, addr.c_str(), nullptr, 0);
-        MAPF_WARN_IF(rc, "nng_dial failed: " << nng_strerror(rc));
+        MAPF_INFO_IF(rc, "nng_dial failed: " << nng_strerror(rc) << " retrying...");
+        if (retry > 0) {
```

Comment:
Shorter option:
```suggestion
        do {
            rc = nng_dial(sock->sd_, addr.c_str(), nullptr, 0);
            MAPF_DEBUG_IF(rc, "nng_dial failed: " << nng_strerror(rc) << " retry #" << retries);
         } while (--retry && rc);
         MAPF_ERROR_IF(rc, "nng_dial_failed, retries #" << retries);
         return rc;
```
Now we don't need to print INFO for failure in the loop, keep it DEBUG, and print ERROR if failed after all retries. No need to print if succeeded.
infinit retries supported with retries<1. One shot - retry=1. All good

Reply:
With that suggestion, retry == 0 will not give infinite retries, but only 2*MAXINT. That's probably enough though. It also gives signed overflow undefined behaviour but we should never get there. 

Still, I'm not too enthusiastic about this suggestion. It seems more difficult to understand.